// @bendera/vscode-webview-elements/dist/vscode-checkbox/vscode-checkbox.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, nothing } from 'lit';
import { customElement, property, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { FormButtonWidgetBase } from '../includes/form-button-widget/FormButtonWidgetBase.js';
import { LabelledCheckboxOrRadioMixin } from '../includes/form-button-widget/LabelledCheckboxOrRadio.js';
import styles from './vscode-checkbox.styles.js';
/**
 * @attr name - Name which is used as a variable name in the data of the form-container.
 * @attr label - Attribute pair of the `label` property.
 * @prop label - Label text. It is only applied if componnet's innerHTML doesn't contain any text.
 *
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vsc-foreground-translucent - Label font color. 90% transparency version of `--vscode-foreground` by default.
 * @cssprop --vscode-settings-checkboxBackground
 * @cssprop --vscode-settings-checkboxBorder
 * @cssprop --vscode-settings-checkboxForeground
 * @cssprop --vscode-focusBorder
 */
export let VscodeCheckbox = class VscodeCheckbox extends LabelledCheckboxOrRadioMixin(FormButtonWidgetBase) {
    constructor() {
        super(...arguments);
        this.role = 'checkbox';
        this.value = '';
        this.disabled = false;
        this._checked = false;
        this._indeterminate = false;
    }
    set checked(val) {
        this._checked = val;
        this._indeterminate = false;
        this.setAttribute('aria-checked', val ? 'true' : 'false');
    }
    get checked() {
        return this._checked;
    }
    set indeterminate(val) {
        this._indeterminate = val;
    }
    get indeterminate() {
        return this._indeterminate;
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('aria-checked', this._checked ? 'true' : 'false');
    }
    _handleClick() {
        if (this.disabled) {
            return;
        }
        this._checked = !this._checked;
        this._indeterminate = false;
        this.setAttribute('aria-checked', this._checked ? 'true' : 'false');
        this.dispatchEvent(new CustomEvent('vsc-change', {
            detail: {
                checked: this._checked,
                label: this.label,
                value: this.value,
            },
            bubbles: true,
            composed: true,
        }));
    }
    _handleKeyDown(ev) {
        if (!this.disabled && (ev.key === 'Enter' || ev.key === ' ')) {
            ev.preventDefault();
            this._checked = !this._checked;
            this.setAttribute('aria-checked', this._checked ? 'true' : 'false');
        }
    }
    render() {
        const iconClasses = classMap({
            icon: true,
            checked: this._checked,
            indeterminate: this._indeterminate,
        });
        const labelInnerClasses = classMap({
            'label-inner': true,
        });
        const icon = html `<svg
      width="16"
      height="16"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
      fill="currentColor"
    >
      <path
        fill-rule="evenodd"
        clip-rule="evenodd"
        d="M14.431 3.323l-8.47 10-.79-.036-3.35-4.77.818-.574 2.978 4.24 8.051-9.506.764.646z"
      />
    </svg>`;
        const check = this._checked && !this._indeterminate ? icon : nothing;
        const indeterminate = this._indeterminate
            ? html `<span class="indeterminate-icon"></span>`
            : nothing;
        return html `
      <div class="wrapper">
        <input
          id="${this._uid}"
          class="checkbox"
          type="checkbox"
          ?checked="${this._checked}"
          value="${this.value}"
          tabindex="-1"
        />
        <div class="${iconClasses}">${indeterminate}${check}</div>
        <label for="${this._uid}" class="label" @click="${this._handleClick}">
          <span class="${labelInnerClasses}">
            ${this._renderLabelAttribute()}
            <slot @slotchange="${this._handleSlotChange}"></slot>
          </span>
        </label>
      </div>
    `;
    }
};
VscodeCheckbox.styles = styles;
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeCheckbox.prototype, "checked", null);
__decorate([
    property({ reflect: true })
], VscodeCheckbox.prototype, "role", void 0);
__decorate([
    property()
], VscodeCheckbox.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeCheckbox.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeCheckbox.prototype, "indeterminate", null);
__decorate([
    state()
], VscodeCheckbox.prototype, "_checked", void 0);
__decorate([
    state()
], VscodeCheckbox.prototype, "_indeterminate", void 0);
VscodeCheckbox = __decorate([
    customElement('vscode-checkbox')
], VscodeCheckbox);
