// @bendera/vscode-webview-elements/dist/vscode-radio/vscode-radio.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { customElement, property, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { FormButtonWidgetBase } from '../includes/form-button-widget/FormButtonWidgetBase.js';
import { LabelledCheckboxOrRadioMixin } from '../includes/form-button-widget/LabelledCheckboxOrRadio.js';
import styles from './vscode-radio.styles.js';
/**
 * @attr name - Name which is used as a variable name in the data of the form-container.
 * @attr label - Attribute pair of the `label` property.
 * @prop label - Label text. It is only applied if component's innerHTML doesn't contain any text.
 *
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vsc-foreground-translucent - Label font color. 90% transparency version of `--vscode-foreground` by default.
 * @cssprop --vscode-settings-checkboxBackground
 * @cssprop --vscode-settings-checkboxBorder
 * @cssprop --vscode-settings-checkboxForeground
 * @cssprop --vscode-focusBorder
 */
export let VscodeRadio = class VscodeRadio extends LabelledCheckboxOrRadioMixin(FormButtonWidgetBase) {
    constructor() {
        super(...arguments);
        /**
         * Name which is used as a variable name in the data of the form-container.
         */
        this.name = '';
        this.value = '';
        this.disabled = false;
        this.role = 'radio';
        this._checked = false;
        this._slottedText = '';
    }
    set checked(val) {
        this._checked = val;
        this.setAttribute('aria-checked', val ? 'true' : 'false');
    }
    get checked() {
        return this._checked;
    }
    _dispatchCustomEvent() {
        this.dispatchEvent(new CustomEvent('vsc-change', {
            detail: {
                checked: this.checked,
                label: this.label,
                value: this.value,
            },
            bubbles: true,
            composed: true,
        }));
    }
    _checkButton() {
        const root = this.getRootNode({ composed: true });
        if (!root) {
            return;
        }
        const radios = root.querySelectorAll(`vscode-radio[name="${this.name}"]`);
        this._checked = true;
        this.setAttribute('aria-checked', 'true');
        radios.forEach((r) => {
            if (r !== this) {
                r.checked = false;
            }
        });
    }
    _handleClick() {
        if (this.disabled) {
            return;
        }
        this._checkButton();
        this._dispatchCustomEvent();
    }
    _handleKeyDown(event) {
        if (!this.disabled && (event.key === 'Enter' || event.key === ' ')) {
            event.preventDefault();
            this._checked = true;
            this.setAttribute('aria-checked', 'true');
            this._dispatchCustomEvent();
        }
    }
    render() {
        const iconClasses = classMap({
            icon: true,
            checked: this._checked,
        });
        const labelInnerClasses = classMap({
            'label-inner': true,
            'is-slot-empty': this._slottedText === '',
        });
        return html `
      <div class="wrapper">
        <input
          id="${this._uid}"
          class="radio"
          type="checkbox"
          ?checked="${this._checked}"
          value="${this.value}"
          tabindex="-1"
        />
        <div class="${iconClasses}"></div>
        <label for="${this._uid}" class="label" @click="${this._handleClick}">
          <span class="${labelInnerClasses}">
            ${this._renderLabelAttribute()}
            <slot @slotchange="${this._handleSlotChange}"></slot>
          </span>
        </label>
      </div>
    `;
    }
};
VscodeRadio.styles = styles;
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeRadio.prototype, "checked", null);
__decorate([
    property()
], VscodeRadio.prototype, "name", void 0);
__decorate([
    property()
], VscodeRadio.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeRadio.prototype, "disabled", void 0);
__decorate([
    property({ reflect: true })
], VscodeRadio.prototype, "role", void 0);
__decorate([
    state()
], VscodeRadio.prototype, "_checked", void 0);
__decorate([
    state()
], VscodeRadio.prototype, "_slottedText", void 0);
VscodeRadio = __decorate([
    customElement('vscode-radio')
], VscodeRadio);
