// @bendera/vscode-webview-elements/dist/vscode-radio-group/vscode-radio-group.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { customElement, property, queryAssignedElements, state, } from 'lit/decorators.js';
import { VscElement } from '../includes/VscElement.js';
import styles from './vscode-radio-group.styles.js';
export let VscodeRadioGroup = class VscodeRadioGroup extends VscElement {
    constructor() {
        super(...arguments);
        this.variant = 'horizontal';
        this.inline = false;
        this.role = 'radiogroup';
        this._focusedRadio = 0;
        this._checkedRadio = -1;
        this._onKeyDownBound = this._onKeyDown.bind(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('keydown', this._onKeyDownBound);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('keydown', this._onKeyDownBound);
    }
    _beforeCheck() {
        const prevChecked = this._checkedRadio;
        const prevFocused = this._focusedRadio;
        if (prevChecked !== -1) {
            this._radios[prevChecked].checked = false;
        }
        if (prevFocused !== -1) {
            this._radios[prevFocused].tabIndex = -1;
        }
    }
    _afterCheck() {
        this._focusedRadio = this._checkedRadio;
        this._radios[this._checkedRadio].checked = true;
        this._radios[this._checkedRadio].tabIndex = 0;
        this._radios[this._checkedRadio].focus();
    }
    _checkPrev() {
        this._beforeCheck();
        if (this._checkedRadio === -1) {
            this._checkedRadio = 0;
        }
        else if (this._checkedRadio - 1 >= 0) {
            this._checkedRadio -= 1;
        }
        else {
            this._checkedRadio = this._radios.length - 1;
        }
        this._afterCheck();
    }
    _checkNext() {
        this._beforeCheck();
        if (this._checkedRadio === -1) {
            this._checkedRadio = 0;
        }
        else if (this._checkedRadio + 1 < this._radios.length) {
            this._checkedRadio += 1;
        }
        else {
            this._checkedRadio = 0;
        }
        this._afterCheck();
    }
    _onKeyDown(ev) {
        const { key } = ev;
        const listenedKeys = ['ArrowLeft', 'ArrowUp', 'ArrowRight', 'ArrowDown'];
        if (listenedKeys.includes(key)) {
            ev.preventDefault();
        }
        if (key === 'ArrowRight' || key === 'ArrowDown') {
            this._checkNext();
        }
        if (key === 'ArrowLeft' || key === 'ArrowUp') {
            this._checkPrev();
        }
    }
    _onChange(ev) {
        const clickedIndex = this._radios.findIndex((r) => r === ev.target);
        if (clickedIndex !== -1) {
            if (this._focusedRadio !== -1) {
                this._radios[this._focusedRadio].tabIndex = -1;
            }
            if (this._checkedRadio !== -1 && this._checkedRadio !== clickedIndex) {
                this._radios[this._checkedRadio].checked = false;
            }
            this._focusedRadio = clickedIndex;
            this._checkedRadio = clickedIndex;
            this._radios[clickedIndex].tabIndex = 0;
        }
    }
    _onSlotChange() {
        this._radios.forEach((r, i) => (r.tabIndex = i === this._focusedRadio ? 0 : -1));
    }
    render() {
        return html `
      <div class="wrapper">
        <slot
          @slotchange=${this._onSlotChange}
          @vsc-change=${this._onChange}
        ></slot>
      </div>
    `;
    }
};
VscodeRadioGroup.styles = styles;
__decorate([
    property({ reflect: true })
], VscodeRadioGroup.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeRadioGroup.prototype, "inline", void 0);
__decorate([
    property({ reflect: true })
], VscodeRadioGroup.prototype, "role", void 0);
__decorate([
    queryAssignedElements({ selector: 'vscode-radio' })
], VscodeRadioGroup.prototype, "_radios", void 0);
__decorate([
    state()
], VscodeRadioGroup.prototype, "_focusedRadio", void 0);
__decorate([
    state()
], VscodeRadioGroup.prototype, "_checkedRadio", void 0);
VscodeRadioGroup = __decorate([
    customElement('vscode-radio-group')
], VscodeRadioGroup);
