// @bendera/vscode-webview-elements/dist/vscode-inputbox/vscode-inputbox.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { customElement, property, query, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { styleMap } from 'lit/directives/style-map.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { VscElement } from '../includes/VscElement.js';
import styles from './vscode-inputbox.styles.js';
var Severity;
(function (Severity) {
    Severity["DEFAULT"] = "default";
    Severity["INFO"] = "info";
    Severity["WARNING"] = "warning";
    Severity["ERROR"] = "error";
})(Severity || (Severity = {}));
const LINE_HEIGHT = 17;
const PADDING = 4;
const BORDER_WIDTH = 1;
const calcHeightFromLines = (lines) => {
    return BORDER_WIDTH * 2 + PADDING * 2 + lines * LINE_HEIGHT;
};
/**
 * @deprecated
 *
 * @attr {narrow|wide} variant - The sizes are borrowed from the VSCode settings page. The narrow size is typically used for the numeric values and the wide size for the text.
 * @attr name - Name which is used as a variable name in the data of the form-container.
 *
 * @cssprop --vscode-scrollbarSlider-background
 * @cssprop --vscode-scrollbarSlider-hoverBackground
 * @cssprop --vscode-scrollbarSlider-activeBackground
 * @cssprop --vscode-input-background
 * @cssprop --vscode-settings-textInputBorder
 * @cssprop --vscode-input-foreground
 * @cssprop --vscode-input-placeholderForeground
 * @cssprop --vscode-focusBorder
 * @cssprop --vscode-panelInput-border
 * @cssprop --vscode-focusBorder
 * @cssprop --vscode-inputValidation-infoBackground
 * @cssprop --vscode-inputValidation-infoBorder
 * @cssprop --vscode-inputValidation-warningBackground
 * @cssprop --vscode-inputValidation-warningBorder
 * @cssprop --vscode-inputValidation-errorBackground
 * @cssprop --vscode-inputValidation-errorBorder
 * @cssprop --vscode-editor-background
 */
export let VscodeInputbox = class VscodeInputbox extends VscElement {
    set severity(val) {
        const oldVal = this._severity;
        switch (val) {
            case Severity.INFO:
            case Severity.WARNING:
            case Severity.ERROR:
                this._severity = val;
                break;
            default:
                this._severity = Severity.DEFAULT;
        }
        this.requestUpdate('messageSeverity', oldVal);
    }
    get severity() {
        return this._severity;
    }
    constructor() {
        super();
        this.label = '';
        this.multiline = false;
        this.message = '';
        /**
         * @deprecated
         * @attr panelInput
         */
        this.panelInput = false;
        /**
         * Text-like input types
         * @attr type
         * @type {"color"|"date"|"datetime-local"|"email"|"file"|"month"|"number"|"password"|"tel"|"text"|"time"|"url"|"week"}
         */
        this.type = 'text';
        this.focused = false;
        this.value = '';
        this.placeholder = '';
        this.lines = 2;
        this.maxLines = 5;
        this.min = undefined;
        this.minLength = undefined;
        this.max = undefined;
        this.maxLength = undefined;
        this.multiple = false;
        this.readonly = false;
        this.step = undefined;
        this._textareaHeight = 0;
        this._textareaDefaultCursor = false;
        this.onInputFocus = () => {
            this.focused = true;
        };
        this.onInputBlur = () => {
            this.focused = false;
        };
        this.onInputInput = (event) => {
            const eventTarget = event.target;
            this.value = eventTarget.value;
            this.dispatchEvent(new CustomEvent('vsc-input', {
                detail: eventTarget.value,
                bubbles: true,
                composed: true,
            }));
            this.resizeTextareaIfRequired();
        };
        this.onInputChange = (event) => {
            const eventTarget = event.target;
            this.dispatchEvent(new CustomEvent('vsc-change', {
                detail: eventTarget.value,
                bubbles: true,
                composed: true,
            }));
        };
        this.onTextareaMouseMove = (event) => {
            const br = this.getBoundingClientRect();
            const x = event.clientX;
            const SCROLLBAR_WIDTH = 10;
            this._textareaDefaultCursor =
                x <= br.left + br.width &&
                    x >= br.left + br.width - SCROLLBAR_WIDTH - BORDER_WIDTH * 2;
            this.requestUpdate();
        };
        this.resizeTextareaIfRequired = () => {
            if (!this._measurerEl || !this.multiline) {
                return;
            }
            const { height } = this._measurerEl.getBoundingClientRect();
            if (height === 0) {
                this._textareaHeight = calcHeightFromLines(this.lines);
                this._measurerEl.style.minHeight = `${calcHeightFromLines(this.lines)}px`;
            }
            else {
                this._textareaHeight = height;
            }
        };
        this._severity = Severity.DEFAULT;
    }
    connectedCallback() {
        super.connectedCallback();
        this.resizeTextareaIfRequired();
    }
    updated(changedProperties) {
        if (changedProperties.has('value')) {
            this.resizeTextareaIfRequired();
        }
    }
    get focusElement() {
        return this._inputElement;
    }
    focus() {
        this._inputElement.focus();
    }
    toString() {
        return '[object VscodeInputbox]';
    }
    render() {
        const minHeight = calcHeightFromLines(this.lines);
        const maxHeight = calcHeightFromLines(this.maxLines);
        const measurerStyles = styleMap({
            minHeight: `${minHeight}px`,
            maxHeight: `${maxHeight}px`,
        });
        const textareaStyles = styleMap({
            height: `${this._textareaHeight}px`,
        });
        const containerClasses = classMap({
            container: true,
            severity: this.severity !== Severity.DEFAULT,
            focused: this.focused,
        });
        const measurerContent = this.value
            ? this.value
                .split('\n')
                .map((line) => line ? html `<div>${line}</div>` : html `<div>&nbsp;</div>`)
            : html `&nbsp;`;
        const textarea = html `
      <textarea
        @focus="${this.onInputFocus}"
        @blur="${this.onInputBlur}"
        @input="${this.onInputInput}"
        @change="${this.onInputChange}"
        @mousemove="${this.onTextareaMouseMove}"
        class="${classMap({
            'cursor-default': this._textareaDefaultCursor,
            'input-element': true,
        })}"
        minlength="${ifDefined(this.minLength)}"
        maxlength="${ifDefined(this.maxLength)}"
        placeholder="${this.placeholder}"
        ?readonly="${this.readonly}"
        style="${textareaStyles}"
        .value="${this.value}"
      ></textarea>
      <div class="content-measurer" style="${measurerStyles}">
        ${measurerContent}
      </div>
    `;
        const input = html `
      <input
        type="${this.type}"
        @focus="${this.onInputFocus}"
        @blur="${this.onInputBlur}"
        @input="${this.onInputInput}"
        @change="${this.onInputChange}"
        placeholder="${this.placeholder}"
        min="${ifDefined(this.min)}"
        minlength="${ifDefined(this.minLength)}"
        max="${ifDefined(this.max)}"
        maxlength="${ifDefined(this.maxLength)}"
        ?multiple="${this.multiple}"
        ?readonly="${this.readonly}"
        step="${ifDefined(this.step)}"
        .value="${this.value}"
        class="input-element"
      />
    `;
        const message = html `
      <div class="message ${this.severity}">${this.message}</div>
    `;
        return html `
      <div class="${containerClasses}">
        <div class="helper"><slot name="helper"></slot></div>
        <div class="input-wrapper">
          ${this.multiline ? textarea : input} ${this.message ? message : ''}
        </div>
      </div>
    `;
    }
};
VscodeInputbox.styles = styles;
__decorate([
    property()
], VscodeInputbox.prototype, "label", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeInputbox.prototype, "multiline", void 0);
__decorate([
    property({ type: String })
], VscodeInputbox.prototype, "message", void 0);
__decorate([
    property({ type: String })
], VscodeInputbox.prototype, "severity", null);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeInputbox.prototype, "panelInput", void 0);
__decorate([
    property({ type: String })
], VscodeInputbox.prototype, "type", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeInputbox.prototype, "focused", void 0);
__decorate([
    property({ type: String })
], VscodeInputbox.prototype, "value", void 0);
__decorate([
    property({ type: String })
], VscodeInputbox.prototype, "placeholder", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "lines", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "maxLines", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "min", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "minLength", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "max", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "maxLength", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeInputbox.prototype, "multiple", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeInputbox.prototype, "readonly", void 0);
__decorate([
    property({ type: Number })
], VscodeInputbox.prototype, "step", void 0);
__decorate([
    query('.content-measurer')
], VscodeInputbox.prototype, "_measurerEl", void 0);
__decorate([
    query('.input-element')
], VscodeInputbox.prototype, "_inputElement", void 0);
__decorate([
    state()
], VscodeInputbox.prototype, "_textareaHeight", void 0);
VscodeInputbox = __decorate([
    customElement('vscode-inputbox')
], VscodeInputbox);
