// @bendera/vscode-webview-elements/dist/vscode-tree/vscode-tree.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, nothing } from 'lit';
import { customElement, property, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { styleMap } from 'lit/directives/style-map.js';
import { VscElement } from '../includes/VscElement.js';
import '../vscode-badge/index.js';
import '../vscode-icon/index.js';
import styles from './vscode-tree.styles.js';
const ARROW_OUTER_WIDTH = 30;
const ARROW_ICON_WIDTH = 16;
const CONTENT_PADDING = 3;
const addPath = (tree, prevPath = []) => {
    const nextTree = [];
    tree.forEach((item, index) => {
        const path = [...prevPath, index];
        const nextItem = {
            ...item,
            path,
        };
        if (item.subItems) {
            nextItem.subItems = addPath(item.subItems, path);
        }
        nextTree.push(nextItem);
    });
    return nextTree;
};
const isBranch = (item) => {
    var _a;
    if (item.subItems &&
        Array.isArray(item.subItems) &&
        ((_a = item === null || item === void 0 ? void 0 : item.subItems) === null || _a === void 0 ? void 0 : _a.length) > 0) {
        return true;
    }
    return false;
};
/**
 * @fires vsc-select Dispatched when an item is selected. The event data shape is described in the
 * `SelectEventDetail` interface.
 * @fires vsc-run-action Dispatched when an action icon is clicked.
 *
 * @cssprop --vscode-focusBorder
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vscode-list-hoverForeground
 * @cssprop --vscode-list-hoverBackground
 * @cssprop --vscode-list-inactiveSelectionBackground
 * @cssprop --vscode-list-inactiveSelectionForeground
 * @cssprop --vscode-list-activeSelectionBackground
 * @cssprop --vscode-list-activeSelectionForeground
 * @cssprop --vscode-list-inactiveSelectionIconForeground
 * @cssprop --vscode-list-inactiveFocusBackground
 * @cssprop --vscode-list-inactiveFocusOutline
 * @cssprop --vscode-list-focusOutline
 * @cssprop --vscode-tree-inactiveIndentGuidesStroke
 * @cssprop --vscode-tree-indentGuidesStroke
 */
export let VscodeTree = class VscodeTree extends VscElement {
    constructor() {
        super(...arguments);
        this.indent = 8;
        this.arrows = false;
        this.multiline = false;
        this.tabindex = 0;
        this.indentGuides = false;
        this._data = [];
        this._selectedItem = null;
        this._focusedItem = null;
        this._selectedBranch = null;
        this._focusedBranch = null;
        this._handleComponentKeyDownBound = this._handleComponentKeyDown.bind(this);
    }
    set data(val) {
        const oldVal = this._data;
        this._data = addPath(val);
        this.requestUpdate('data', oldVal);
    }
    get data() {
        return this._data;
    }
    closeAll() {
        this._closeSubTreeRecursively(this.data);
        this.requestUpdate();
    }
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('keydown', this._handleComponentKeyDownBound);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('keydown', this._handleComponentKeyDownBound);
    }
    _getItemByPath(path) {
        let current = this._data;
        let item = null;
        path.forEach((el, i) => {
            if (i === path.length - 1) {
                item = current[el];
            }
            else {
                current = current[el].subItems;
            }
        });
        return item;
    }
    _handleActionClick(ev) {
        ev.stopPropagation();
        const el = ev.target;
        const itemPath = el.dataset.itemPath;
        const actionIndex = el.dataset.index;
        let item = null;
        let actionId = '';
        let value = '';
        if (itemPath) {
            const path = itemPath.split('/').map((p) => Number(p));
            item = this._getItemByPath(path);
            if (item === null || item === void 0 ? void 0 : item.actions) {
                const index = Number(actionIndex);
                if (item.actions[index]) {
                    actionId = item.actions[index].actionId;
                }
            }
            if (item === null || item === void 0 ? void 0 : item.value) {
                value = item.value;
            }
        }
        this.dispatchEvent(new CustomEvent('vsc-run-action', {
            detail: {
                actionId,
                item,
                value,
            },
        }));
    }
    _renderIconVariant(variant) {
        const { type, value } = variant;
        if (type === 'themeicon') {
            return html `<vscode-icon name=${value} class="theme-icon"></vscode-icon>`;
        }
        else {
            return html `<span
        class="image-icon"
        style="background-image: url(${value});"
      ></span>`;
        }
    }
    _renderIcon(item) {
        const iconVariants = {
            branch: {
                value: 'folder',
                type: 'themeicon',
            },
            open: {
                value: 'folder-opened',
                type: 'themeicon',
            },
            leaf: {
                value: 'file',
                type: 'themeicon',
            },
        };
        if (item.iconUrls) {
            if (item.iconUrls.branch) {
                iconVariants.branch = {
                    value: item.iconUrls.branch,
                    type: 'image',
                };
            }
            if (item.iconUrls.leaf) {
                iconVariants.leaf = {
                    value: item.iconUrls.leaf,
                    type: 'image',
                };
            }
            if (item.iconUrls.open) {
                iconVariants.open = {
                    value: item.iconUrls.open,
                    type: 'image',
                };
            }
        }
        else if (typeof item.icons === 'object') {
            if (item.icons.branch) {
                iconVariants.branch = {
                    value: item.icons.branch,
                    type: 'themeicon',
                };
            }
            if (item.icons.leaf) {
                iconVariants.leaf = {
                    value: item.icons.leaf,
                    type: 'themeicon',
                };
            }
            if (item.icons.open) {
                iconVariants.open = {
                    value: item.icons.open,
                    type: 'themeicon',
                };
            }
        }
        else if (!item.icons) {
            return html `${nothing}`;
        }
        if (isBranch(item)) {
            if (item.open) {
                return this._renderIconVariant(iconVariants.open);
            }
            else {
                return this._renderIconVariant(iconVariants.branch);
            }
        }
        else {
            return this._renderIconVariant(iconVariants.leaf);
        }
    }
    _renderArrow(item) {
        if (!this.arrows || !isBranch(item)) {
            return html `${nothing}`;
        }
        const { open = false } = item;
        const arrowIconName = open ? 'chevron-down' : 'chevron-right';
        return html `
      <div class="arrow-container">
        <vscode-icon name="${arrowIconName}" class="icon-arrow"></vscode-icon>
      </div>
    `;
    }
    _renderActions(item) {
        const actionButtons = [];
        if (item.actions && Array.isArray(item.actions)) {
            item.actions.forEach((action, index) => {
                var _a;
                if (action.icon) {
                    const icon = html `<vscode-icon
            name=${action.icon}
            action-icon
            title=${ifDefined(action.tooltip)}
            data-item-path=${ifDefined((_a = item.path) === null || _a === void 0 ? void 0 : _a.join('/'))}
            data-index=${index}
            class="action-icon"
            @click=${this._handleActionClick}
          ></vscode-icon>`;
                    actionButtons.push(icon);
                }
            });
        }
        if (actionButtons.length > 0) {
            return html `<div class="actions">${actionButtons}</div>`;
        }
        else {
            return html `${nothing}`;
        }
    }
    _renderDecorations(item) {
        const decorations = [];
        if (item.decorations && Array.isArray(item.decorations)) {
            item.decorations.forEach((decoration) => {
                const { appearance = 'text', visibleWhen = 'always', content = '', color = '', focusedColor = '', hoverColor = '', selectedColor = '', } = decoration;
                const visibleWhenClass = `visible-when-${visibleWhen}`;
                const inlineStyles = {};
                if (color) {
                    inlineStyles['--color'] = color;
                }
                if (focusedColor) {
                    inlineStyles['--focused-color'] = focusedColor;
                }
                if (hoverColor) {
                    inlineStyles['--hover-color'] = hoverColor;
                }
                if (selectedColor) {
                    inlineStyles['--selected-color'] = selectedColor;
                }
                switch (appearance) {
                    case 'counter-badge':
                        decorations.push(html `<vscode-badge
                variant="counter"
                class=${['counter-badge', visibleWhenClass].join(' ')}
                >${content}</vscode-badge
              >`);
                        break;
                    case 'filled-circle':
                        decorations.push(html `<vscode-icon
                name="circle-filled"
                size="14"
                class=${['filled-circle', visibleWhenClass].join(' ')}
                style=${styleMap(inlineStyles)}
              ></vscode-icon>`);
                        break;
                    case 'text':
                        decorations.push(html `<div
                class=${['decoration-text', visibleWhenClass].join(' ')}
                style=${styleMap(inlineStyles)}
              >
                ${content}
              </div>`);
                        break;
                    default:
                }
            });
        }
        if (decorations.length > 0) {
            return html `<div class="decorations">${decorations}</div>`;
        }
        else {
            return html `${nothing}`;
        }
    }
    _renderTreeItem(item, additionalOptions) {
        const { open = false, label, description = '', selected = false, focused = false, subItems = [], } = item;
        const { path, itemType, hasFocusedItem = false, hasSelectedItem = false, } = additionalOptions;
        const indentLevel = path.length - 1;
        const contentsClasses = ['contents'];
        const liClasses = open ? ['open'] : [];
        const indentSize = indentLevel * this.indent;
        const padLeft = this.arrows && itemType === 'leaf'
            ? ARROW_OUTER_WIDTH + indentSize
            : indentSize;
        const arrowMarkup = this._renderArrow(item);
        const iconMarkup = this._renderIcon(item);
        const indentGuidePos = this.arrows
            ? indentSize + ARROW_ICON_WIDTH
            : indentSize + CONTENT_PADDING;
        const subTreeMarkup = open && itemType === 'branch'
            ? html `<ul
            style=${styleMap({ '--indent-guide-pos': `${indentGuidePos}px` })}
            class=${classMap({
                'has-active-item': hasFocusedItem || hasSelectedItem,
            })}
          >
            ${this._renderTree(subItems, path)}
          </ul>`
            : nothing;
        const descriptionMarkup = description
            ? html `<span class="description">${description}</span>`
            : nothing;
        const actionsMarkup = this._renderActions(item);
        const decorationsMarkup = this._renderDecorations(item);
        liClasses.push(itemType);
        if (selected) {
            contentsClasses.push('selected');
        }
        if (focused) {
            contentsClasses.push('focused');
        }
        return html `
      <li data-path="${path.join('/')}" class="${liClasses.join(' ')}">
        <div
          class="${contentsClasses.join(' ')}"
          style="${styleMap({ paddingLeft: `${padLeft + CONTENT_PADDING}px` })}"
        >
          ${arrowMarkup}${iconMarkup}<span class="text-content"
            >${label}${descriptionMarkup}</span
          >
          ${actionsMarkup} ${decorationsMarkup}
        </div>
        ${subTreeMarkup}
      </li>
    `;
    }
    _renderTree(tree, oldPath = []) {
        const ret = [];
        if (!tree) {
            return nothing;
        }
        tree.forEach((item, index) => {
            const path = [...oldPath, index];
            const itemType = isBranch(item) ? 'branch' : 'leaf';
            const { selected = false, focused = false, hasFocusedItem = false, hasSelectedItem = false, } = item;
            if (selected) {
                this._selectedItem = item;
            }
            if (focused) {
                this._focusedItem = item;
            }
            ret.push(this._renderTreeItem(item, {
                path,
                itemType,
                hasFocusedItem,
                hasSelectedItem,
            }));
        });
        return ret;
    }
    _selectItem(item) {
        var _a, _b;
        if (this._selectedItem) {
            this._selectedItem.selected = false;
        }
        if (this._focusedItem) {
            this._focusedItem.focused = false;
        }
        this._selectedItem = item;
        item.selected = true;
        this._focusedItem = item;
        item.focused = true;
        if (this._selectedBranch) {
            this._selectedBranch.hasSelectedItem = false;
        }
        let parentBranch = null;
        if (((_a = item.path) === null || _a === void 0 ? void 0 : _a.length) && item.path.length > 1) {
            parentBranch = this._getItemByPath(item.path.slice(0, -1));
        }
        if (isBranch(item)) {
            this._selectedBranch = item;
            item.hasSelectedItem = true;
            item.open = !item.open;
            if (!item.open) {
                if (parentBranch) {
                    this._selectedBranch = parentBranch;
                    parentBranch.hasSelectedItem = true;
                }
            }
            else {
                this._selectedBranch = item;
                item.hasSelectedItem = true;
            }
        }
        else {
            if (((_b = item.path) === null || _b === void 0 ? void 0 : _b.length) && item.path.length > 1) {
                const parentBranch = this._getItemByPath(item.path.slice(0, -1));
                if (parentBranch) {
                    this._selectedBranch = parentBranch;
                    parentBranch.hasSelectedItem = true;
                }
            }
            else {
                this._selectedBranch = item;
                item.hasSelectedItem = true;
            }
        }
        this._emitSelectEvent(this._selectedItem, this._selectedItem.path.join('/'));
        this.requestUpdate();
    }
    _focusItem(item) {
        var _a, _b;
        if (this._focusedItem) {
            this._focusedItem.focused = false;
        }
        this._focusedItem = item;
        item.focused = true;
        const isBranch = !!((_a = item === null || item === void 0 ? void 0 : item.subItems) === null || _a === void 0 ? void 0 : _a.length);
        if (this._focusedBranch) {
            this._focusedBranch.hasFocusedItem = false;
        }
        let parentBranch = null;
        if (((_b = item.path) === null || _b === void 0 ? void 0 : _b.length) && item.path.length > 1) {
            parentBranch = this._getItemByPath(item.path.slice(0, -1));
        }
        if (!isBranch) {
            if (parentBranch) {
                this._focusedBranch = parentBranch;
                parentBranch.hasFocusedItem = true;
            }
        }
        else {
            if (item.open) {
                this._focusedBranch = item;
                item.hasFocusedItem = true;
            }
            else if (!item.open && parentBranch) {
                this._focusedBranch = parentBranch;
                parentBranch.hasFocusedItem = true;
            }
        }
    }
    _closeSubTreeRecursively(tree) {
        tree.forEach((item) => {
            item.open = false;
            if (item.subItems && item.subItems.length > 0) {
                this._closeSubTreeRecursively(item.subItems);
            }
        });
    }
    _emitSelectEvent(item, path) {
        const { icons, label, open, value } = item;
        const detail = {
            icons,
            itemType: isBranch(item) ? 'branch' : 'leaf',
            label,
            open: open || false,
            value: value || label,
            path,
        };
        this.dispatchEvent(new CustomEvent('vsc-select', {
            bubbles: true,
            composed: true,
            detail,
        }));
    }
    _focusPrevItem() {
        var _a;
        if (!this._focusedItem) {
            this._focusItem(this._data[0]);
            return;
        }
        const { path } = this._focusedItem;
        if (path && (path === null || path === void 0 ? void 0 : path.length) > 0) {
            const currentItemIndex = path[path.length - 1];
            const hasParent = path.length > 1;
            if (currentItemIndex > 0) {
                const newPath = [...path];
                newPath[newPath.length - 1] = currentItemIndex - 1;
                const prevSibling = this._getItemByPath(newPath);
                let newFocusedItem = prevSibling;
                if ((prevSibling === null || prevSibling === void 0 ? void 0 : prevSibling.open) && ((_a = prevSibling.subItems) === null || _a === void 0 ? void 0 : _a.length)) {
                    const { subItems } = prevSibling;
                    newFocusedItem = subItems[subItems.length - 1];
                }
                this._focusItem(newFocusedItem);
            }
            else {
                if (hasParent) {
                    const newPath = [...path];
                    newPath.pop();
                    this._focusItem(this._getItemByPath(newPath));
                }
            }
        }
        else {
            this._focusItem(this._data[0]);
        }
    }
    _focusNextItem() {
        if (!this._focusedItem) {
            this._focusItem(this._data[0]);
            return;
        }
        const { path, open } = this._focusedItem;
        if (open &&
            Array.isArray(this._focusedItem.subItems) &&
            this._focusedItem.subItems.length > 0) {
            this._focusItem(this._focusedItem.subItems[0]);
            return;
        }
        const nextPath = [...path];
        nextPath[nextPath.length - 1] += 1;
        let nextFocusedItem = this._getItemByPath(nextPath);
        if (nextFocusedItem) {
            this._focusItem(nextFocusedItem);
        }
        else {
            nextPath.pop();
            if (nextPath.length > 0) {
                nextPath[nextPath.length - 1] += 1;
                nextFocusedItem = this._getItemByPath(nextPath);
                if (nextFocusedItem) {
                    this._focusItem(nextFocusedItem);
                }
            }
        }
    }
    _handleClick(event) {
        const composedPath = event.composedPath();
        const targetElement = composedPath.find((el) => el.tagName &&
            el.tagName.toUpperCase() === 'LI');
        if (targetElement) {
            const pathStr = targetElement.dataset.path || '';
            const path = pathStr.split('/').map((el) => Number(el));
            const item = this._getItemByPath(path);
            this._selectItem(item);
        }
        else {
            if (this._focusedItem) {
                this._focusedItem.focused = false;
            }
            this._focusedItem = null;
        }
    }
    _handleComponentKeyDown(ev) {
        const keys = [
            ' ',
            'ArrowDown',
            'ArrowUp',
            'Enter',
            'Escape',
        ];
        const key = ev.key;
        if (keys.includes(ev.key)) {
            ev.stopPropagation();
            ev.preventDefault();
        }
        if (key === 'Escape') {
            this._focusedItem = null;
        }
        if (key === 'ArrowUp') {
            this._focusPrevItem();
        }
        if (key === 'ArrowDown') {
            this._focusNextItem();
        }
        if (key === 'Enter' || key === ' ') {
            if (this._focusedItem) {
                this._selectItem(this._focusedItem);
            }
        }
    }
    render() {
        const classes = classMap({
            multi: this.multiline,
            single: !this.multiline,
            wrapper: true,
            'has-not-focused-item': !this._focusedItem,
            'selection-none': !this._selectedItem,
            'selection-single': this._selectedItem !== null,
        });
        return html `
      <div @click="${this._handleClick}" class="${classes}">
        <ul>
          ${this._renderTree(this._data)}
        </ul>
      </div>
    `;
    }
};
VscodeTree.styles = styles;
__decorate([
    property({ type: Array, reflect: false })
], VscodeTree.prototype, "data", null);
__decorate([
    property({ type: Number })
], VscodeTree.prototype, "indent", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTree.prototype, "arrows", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTree.prototype, "multiline", void 0);
__decorate([
    property({ type: Number, reflect: true })
], VscodeTree.prototype, "tabindex", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'indent-guides' })
], VscodeTree.prototype, "indentGuides", void 0);
__decorate([
    state()
], VscodeTree.prototype, "_selectedItem", void 0);
__decorate([
    state()
], VscodeTree.prototype, "_focusedItem", void 0);
__decorate([
    state()
], VscodeTree.prototype, "_selectedBranch", void 0);
__decorate([
    state()
], VscodeTree.prototype, "_focusedBranch", void 0);
VscodeTree = __decorate([
    customElement('vscode-tree')
], VscodeTree);
