// @bendera/vscode-webview-elements/dist/vscode-textfield/vscode-textfield.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { customElement, property, query } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { VscElement } from '../includes/VscElement.js';
import styles from './vscode-textfield.styles.js';
/**
 * A simple inline textfield
 *
 * @slot content-before
 * @slot content-after
 *
 * @cssprop --vscode-settings-textInputBackground
 * @cssprop --vscode-settings-textInputBorder
 * @cssprop --vscode-settings-textInputForeground
 * @cssprop --vscode-focusBorder
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vscode-input-placeholderForeground
 * @cssprop --vscode-button-background
 * @cssprop --vscode-button-foreground
 * @cssprop --vscode-button-hoverBackground
 */
export let VscodeTextfield = class VscodeTextfield extends VscElement {
    constructor() {
        super(...arguments);
        this.autocomplete = undefined;
        this.autofocus = false;
        this.disabled = false;
        this.focused = false;
        this.invalid = false;
        /**
         * Set `aria-label` for the inner input element. Should not be set,
         * vscode-label will do it automatically.
         */
        this.label = '';
        this.max = undefined;
        this.maxlength = undefined;
        this.min = undefined;
        this.minlength = undefined;
        this.multiple = false;
        this.name = undefined;
        this.pattern = undefined;
        this.placeholder = undefined;
        this.readonly = false;
        this.required = false;
        this.step = undefined;
        this.type = 'text';
        this._value = '';
    }
    set value(val) {
        this._value = val;
    }
    get value() {
        return this._value;
    }
    get wrappedElement() {
        return this._inputEl;
    }
    connectedCallback() {
        super.connectedCallback();
        this.updateComplete.then(() => {
            this._validate();
        });
    }
    focus() {
        this._inputEl.focus();
    }
    checkValidity() {
        this._validate();
        return !this.invalid;
    }
    _validate() {
        this.invalid = !this._inputEl.checkValidity();
    }
    _onInvalid(ev) {
        this.dispatchEvent(new CustomEvent('vsc-invalid', { detail: { originalEvent: ev } }));
    }
    _onInput(ev) {
        this._value = this._inputEl.value;
        this.dispatchEvent(new CustomEvent('vsc-input', { detail: { data: ev.data, originalEvent: ev } }));
    }
    _onChange(ev) {
        this._value = this._inputEl.value;
        this.dispatchEvent(new CustomEvent('vsc-change', {
            detail: { data: ev.data, originalEvent: ev },
        }));
    }
    _onFocus() {
        this.focused = true;
    }
    _onBlur() {
        this.focused = false;
    }
    render() {
        return html `
      <slot name="content-before"></slot>
      <input
        id="input"
        type=${this.type}
        ?autofocus=${this.autofocus}
        autocomplete=${ifDefined(this.autocomplete)}
        aria-label=${this.label}
        ?disabled=${this.disabled}
        max=${ifDefined(this.max)}
        maxlength=${ifDefined(this.maxlength)}
        min=${ifDefined(this.min)}
        minlength=${ifDefined(this.minlength)}
        ?multiple=${this.multiple}
        name=${ifDefined(this.name)}
        pattern=${ifDefined(this.pattern)}
        placeholder=${ifDefined(this.placeholder)}
        ?readonly=${this.readonly}
        ?required=${this.required}
        step=${ifDefined(this.step)}
        .value=${this._value}
        @blur=${this._onBlur}
        @change=${this._onChange}
        @focus=${this._onFocus}
        @input=${this._onInput}
        @invalid=${this._onInvalid}
      />
      <slot name="content-after"></slot>
    `;
    }
};
VscodeTextfield.styles = styles;
__decorate([
    property()
], VscodeTextfield.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean })
], VscodeTextfield.prototype, "autofocus", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextfield.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextfield.prototype, "focused", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextfield.prototype, "invalid", void 0);
__decorate([
    property({ attribute: false })
], VscodeTextfield.prototype, "label", void 0);
__decorate([
    property({ type: Number })
], VscodeTextfield.prototype, "max", void 0);
__decorate([
    property({ type: Number })
], VscodeTextfield.prototype, "maxlength", void 0);
__decorate([
    property({ type: Number })
], VscodeTextfield.prototype, "min", void 0);
__decorate([
    property({ type: Number })
], VscodeTextfield.prototype, "minlength", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextfield.prototype, "multiple", void 0);
__decorate([
    property()
], VscodeTextfield.prototype, "name", void 0);
__decorate([
    property()
], VscodeTextfield.prototype, "pattern", void 0);
__decorate([
    property()
], VscodeTextfield.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextfield.prototype, "readonly", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextfield.prototype, "required", void 0);
__decorate([
    property({ type: Number })
], VscodeTextfield.prototype, "step", void 0);
__decorate([
    property()
], VscodeTextfield.prototype, "type", void 0);
__decorate([
    property()
], VscodeTextfield.prototype, "value", null);
__decorate([
    query('#input')
], VscodeTextfield.prototype, "_inputEl", void 0);
VscodeTextfield = __decorate([
    customElement('vscode-textfield')
], VscodeTextfield);
