// @bendera/vscode-webview-elements/dist/vscode-button/vscode-button.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, nothing } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { VscElement } from '../includes/VscElement.js';
import '../vscode-icon/index.js';
import styles from './vscode-button.styles.js';
/**
 * @fires vsc-click Dispatched only when button is not in disabled state.
 *
 * @cssprop --vscode-button-background
 * @cssprop --vscode-button-foreground
 * @cssprop [--vscode-button-border=var(--vscode-button-background)]
 * @cssprop --vscode-button-hoverBackground
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vscode-button-secondaryForeground
 * @cssprop --vscode-button-secondaryBackground
 * @cssprop --vscode-button-secondaryHoverBackground
 * @cssprop --vscode-focusBorder
 */
export let VscodeButton = class VscodeButton extends VscElement {
    constructor() {
        super();
        this.tabindex = 0;
        /**
         * Button has a less prominent style.
         */
        this.secondary = false;
        this.role = 'button';
        this.disabled = false;
        /**
         * A [Codicon](https://microsoft.github.io/vscode-codicons/dist/codicon.html) before the label
         */
        this.icon = '';
        /**
         * A [Codicon](https://microsoft.github.io/vscode-codicons/dist/codicon.html) after the label
         */
        this.iconAfter = '';
        this.focused = false;
        this._prevTabindex = 0;
        this._handleFocusBound = this._handleFocus.bind(this);
        this._handleBlurBound = this._handleBlur.bind(this);
        this.addEventListener('keydown', this._handleKeyDown.bind(this));
        this.addEventListener('click', this._handleClick.bind(this));
    }
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('focus', this._handleFocusBound);
        this.addEventListener('blur', this._handleBlurBound);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('focus', this._handleFocusBound);
        this.removeEventListener('blur', this._handleBlurBound);
    }
    attributeChangedCallback(name, oldVal, newVal) {
        super.attributeChangedCallback(name, oldVal, newVal);
        if (name === 'disabled' && this.hasAttribute('disabled')) {
            this._prevTabindex = this.tabindex;
            this.tabindex = -1;
        }
        else if (name === 'disabled' && !this.hasAttribute('disabled')) {
            this.tabindex = this._prevTabindex;
        }
    }
    _handleKeyDown(event) {
        if ((event.key === 'Enter' || event.key === ' ') &&
            !this.hasAttribute('disabled')) {
            this.dispatchEvent(new CustomEvent('vsc-click', {
                detail: {
                    originalEvent: new MouseEvent('click'),
                },
            }));
        }
    }
    _handleClick(event) {
        if (!this.hasAttribute('disabled')) {
            this.dispatchEvent(new CustomEvent('vsc-click', {
                detail: {
                    originalEvent: event,
                },
            }));
        }
    }
    _handleFocus() {
        this.focused = true;
    }
    _handleBlur() {
        this.focused = false;
    }
    render() {
        const hasIcon = this.icon !== '';
        const hasIconAfter = this.iconAfter !== '';
        const wrapperClasses = {
            wrapper: true,
            'has-icon-before': hasIcon,
            'has-icon-after': hasIconAfter,
        };
        const iconElem = hasIcon
            ? html `<vscode-icon name="${this.icon}" class="icon"></vscode-icon>`
            : nothing;
        const iconAfterElem = hasIconAfter
            ? html `<vscode-icon
          name="${this.iconAfter}"
          class="icon-after"
        ></vscode-icon>`
            : nothing;
        return html `
      <span class="${classMap(wrapperClasses)}">
        ${iconElem}
        <slot></slot>
        ${iconAfterElem}
      </span>
    `;
    }
};
VscodeButton.styles = styles;
__decorate([
    property({ type: Number, reflect: true })
], VscodeButton.prototype, "tabindex", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "secondary", void 0);
__decorate([
    property({ reflect: true })
], VscodeButton.prototype, "role", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "disabled", void 0);
__decorate([
    property()
], VscodeButton.prototype, "icon", void 0);
__decorate([
    property({ attribute: 'icon-after' })
], VscodeButton.prototype, "iconAfter", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeButton.prototype, "focused", void 0);
VscodeButton = __decorate([
    customElement('vscode-button')
], VscodeButton);
