// @bendera/vscode-webview-elements/dist/vscode-textarea/vscode-textarea.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { customElement, property, query, state } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { classMap } from 'lit/directives/class-map.js';
import { styleMap } from 'lit/directives/style-map.js';
import { VscElement } from '../includes/VscElement.js';
import styles from './vscode-textarea.styles.js';
/**
 * Multi-line text input.
 *
 * @cssprop --vscode-scrollbar-shadow
 * @cssprop --vscode-settings-textInputBackground
 * @cssprop --vscode-settings-textInputBorder
 * @cssprop --vscode-settings-textInputForeground
 * @cssprop --vscode-input-placeholderForeground
 * @cssprop --vscode-font-family
 * @cssprop --vscode-font-size
 * @cssprop --vscode-font-weight
 * @cssprop --vscode-editor-background
 * @cssprop --vscode-editor-foreground
 * @cssprop --vscode-editor-font-family
 * @cssprop --vscode-editor-font-size
 * @cssprop --vscode-editor-font-weight
 * @cssprop --vscode-editor-inlineValuesForeground
 * @cssprop --vscode-focusBorder
 * @cssprop --vscode-scrollbarSlider-background
 * @cssprop --vscode-scrollbarSlider-hoverBackground
 * @cssprop --vscode-scrollbarSlider-activeBackground
 */
export let VscodeTextarea = class VscodeTextarea extends VscElement {
    constructor() {
        super(...arguments);
        this.autocomplete = undefined;
        this.autofocus = false;
        this.disabled = false;
        this.label = '';
        this.maxlength = undefined;
        this.minlength = undefined;
        this.rows = undefined;
        this.cols = undefined;
        this.name = undefined;
        this.placeholder = undefined;
        this.readonly = false;
        this.resize = 'none';
        this.required = false;
        this.spellcheck = false;
        /**
         * Use monospace fonts. The font family, weight, size, and color will be the same as set in the
         * VSCode code editor.
         */
        this.monospace = false;
        this._value = '';
        this._textareaPointerCursor = false;
        this._shadow = false;
    }
    set value(val) {
        this._value = val;
    }
    get value() {
        return this._value;
    }
    /**
     * Getter for the inner textarea element if it needs to be accessed for some reason.
     */
    get wrappedElement() {
        return this._textareaEl;
    }
    focus() {
        this._textareaEl.focus();
    }
    _handleChange(ev) {
        this._value = this._textareaEl.value;
        this.dispatchEvent(new CustomEvent('vsc-change', {
            detail: { data: ev.data, originalEvent: ev },
        }));
    }
    _handleInput(ev) {
        this._value = this._textareaEl.value;
        this.dispatchEvent(new CustomEvent('vsc-input', {
            detail: { data: ev.data, originalEvent: ev },
        }));
    }
    _handleMouseMove(ev) {
        if (this._textareaEl.clientHeight >= this._textareaEl.scrollHeight) {
            this._textareaPointerCursor = false;
            return;
        }
        const SCROLLBAR_WIDTH = 14;
        const BORDER_WIDTH = 1;
        const br = this._textareaEl.getBoundingClientRect();
        const x = ev.clientX;
        this._textareaPointerCursor =
            x >= br.left + br.width - SCROLLBAR_WIDTH - BORDER_WIDTH * 2;
    }
    _handleScroll() {
        this._shadow = this._textareaEl.scrollTop > 0;
    }
    render() {
        return html `
      <div
        class=${classMap({
            shadow: true,
            visible: this._shadow,
        })}
      ></div>
      <textarea
        autocomplete=${ifDefined(this.autocomplete)}
        ?autofocus=${this.autofocus}
        ?disabled=${this.disabled}
        aria-label=${this.label}
        id="textarea"
        class=${classMap({
            monospace: this.monospace,
            'cursor-pointer': this._textareaPointerCursor,
        })}
        maxlength=${ifDefined(this.maxlength)}
        minlength=${ifDefined(this.minlength)}
        rows=${ifDefined(this.rows)}
        cols=${ifDefined(this.cols)}
        name=${ifDefined(this.name)}
        placeholder=${ifDefined(this.placeholder)}
        ?readonly=${this.readonly}
        style=${styleMap({
            resize: this.resize,
        })}
        ?required=${this.required}
        spellcheck=${this.spellcheck}
        @change=${this._handleChange}
        @input=${this._handleInput}
        @mousemove=${this._handleMouseMove}
        @scroll=${this._handleScroll}
      >
${this._value}</textarea
      >
    `;
    }
};
VscodeTextarea.styles = styles;
__decorate([
    property()
], VscodeTextarea.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean })
], VscodeTextarea.prototype, "autofocus", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextarea.prototype, "disabled", void 0);
__decorate([
    property({ attribute: false })
], VscodeTextarea.prototype, "label", void 0);
__decorate([
    property({ type: Number })
], VscodeTextarea.prototype, "maxlength", void 0);
__decorate([
    property({ type: Number })
], VscodeTextarea.prototype, "minlength", void 0);
__decorate([
    property({ type: Number })
], VscodeTextarea.prototype, "rows", void 0);
__decorate([
    property({ type: Number })
], VscodeTextarea.prototype, "cols", void 0);
__decorate([
    property()
], VscodeTextarea.prototype, "name", void 0);
__decorate([
    property()
], VscodeTextarea.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextarea.prototype, "readonly", void 0);
__decorate([
    property()
], VscodeTextarea.prototype, "resize", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextarea.prototype, "required", void 0);
__decorate([
    property({ type: Boolean })
], VscodeTextarea.prototype, "spellcheck", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], VscodeTextarea.prototype, "monospace", void 0);
__decorate([
    property()
], VscodeTextarea.prototype, "value", null);
__decorate([
    query('#textarea')
], VscodeTextarea.prototype, "_textareaEl", void 0);
__decorate([
    state()
], VscodeTextarea.prototype, "_value", void 0);
__decorate([
    state()
], VscodeTextarea.prototype, "_textareaPointerCursor", void 0);
__decorate([
    state()
], VscodeTextarea.prototype, "_shadow", void 0);
VscodeTextarea = __decorate([
    customElement('vscode-textarea')
], VscodeTextarea);
